File: src/main/java/com/redbox/domain/notice/dto/NoticeResponse.java
package com.redbox.domain.notice.dto;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.redbox.domain.attach.dto.AttachFileResponse;
import com.redbox.domain.notice.entity.Notice;
import com.redbox.domain.user.entity.User;
import lombok.Getter;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

@Getter
@JsonDeserialize
public class NoticeResponse {

    private Long noticeNo;
    private String title;
    private String content;
    private LocalDate createdDate;
    private String writer;
    private int views;
    private List<AttachFileResponse> attachFileResponses;

    @JsonCreator
    public NoticeResponse() {
    }

    public NoticeResponse(Notice notice) {
        this.noticeNo = notice.getId();
        this.title = notice.getNoticeTitle();
        this.content = notice.getNoticeContent();
        this.createdDate = notice.getCreatedAt().toLocalDate();
        this.writer = Optional.ofNullable(notice.getUser())
                .map(User::getName)
                .orElse("Unknown");
        this.views = notice.getNoticeHits();
        this.attachFileResponses = notice.getAttachFiles()
                .stream().map(AttachFileResponse::new).toList();

    }
}

File: migration/src/main/kotlin/com/redbox/domain/community/notice/dto/NoticeResponse.kt
package com.redbox.domain.community.notice.dto

import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.redbox.domain.community.attach.dto.AttachFileResponse
import com.redbox.domain.community.notice.entity.Notice
import java.time.LocalDate

@JsonDeserialize
@JsonTypeInfo(use = JsonTypeInfo.Id.CLASS, include = JsonTypeInfo.As.PROPERTY, property = "@class")
data class NoticeResponse(
    val noticeNo: Long,
    val title: String,
    val content: String,
    val createdDate: LocalDate,
    val writer: String,
    val views: Int,
    val attachFileResponses: List<AttachFileResponse>
) {
    companion object {
        fun fromNotice(notice: Notice): NoticeResponse {
            return NoticeResponse(
                noticeNo = notice.id ?: throw IllegalStateException("Notice not found"),
                title = notice.noticeTitle ?: "No Title",
                content = notice.noticeContent ?: "No Content",
                createdDate = notice.createdAt?.toLocalDate() ?: LocalDate.now(),
                writer = notice.user?.name ?: throw IllegalStateException("User not found"),
                views = notice.noticeHits ?: 0,
                attachFileResponses = notice.attachFiles.map { attachFile -> AttachFileResponse(attachFile) }
            )
        }
    }
}